PFCluster truth selector

rafaellopesdesa opened the pull request:
This is a small tool that we use at EGM to select ECAL PFcluster matched to GEN photons that did not convert before the calorimeter. The commits are uploaded in EgammaMCTools because of the dependency on GEN (for matching) and SIM (to veto conversions using trackingParticles).

We need it for 90X in the context of the high-level PF studies for the 2017 run, but we are PRing here first (as it is required) and, after review and integration in 91X, we will immediately backport to 90X.

The PR does not alter any standard validation sequence since it is not run by default. It consist of 4 files: the EDProducer, the BuildFile to compile, the python cfi for the producer and a small test script that basically does what this tool is supposed to used to: make ECAL PF Cluster, make trackingParticles, match and select.

Attn: @jainshilpi , @fcouderc

rafaellopesdesa committed:
select ecal pfclusters matched to unconverted photons

rafaellopesdesa committed:
makes a cleaner test for the filter

cmsbuild commented:
A new Pull Request was created by  @rafaellopesdesa (Rafael Lopes de Sa) for master.

It involves the following packages:

RecoEgamma/EgammaMCTools


@perrotta, @cmsbuild, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@Sam-Harper, @lgray this is something you requested to watch as well.
@muzaffar, @davidlange6, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: bca05dc0b6686a50d6d60e7ea5917a6087dda2c2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/18755/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/18755/summary.html

perrotta commented on RecoEgamma/EgammaMCTools/plugins/PFClusterMatchedToPhotonsSelector.cc:
```diff
@@ -0,0 +1,114 @@
+#include "SimDataFormats/TrackingAnalysis/interface/TrackingParticleFwd.h"
+
+
+class PFClusterMatchedToPhotonsSelector : public edm::EDProducer {
```
Is there any reason for it being a legacy module?
Shouldn't it better inherit from public edm::stream::EDProducer<> ?

perrotta commented on RecoEgamma/EgammaMCTools/plugins/PFClusterMatchedToPhotonsSelector.cc:
```diff
@@ -0,0 +1,114 @@
+
+      bool isConversion = false;
+      for (auto&& vertRef : trackingParticle.decayVertices()) {
+	if (vertRef->position().rho() > 123.8 && std::abs(vertRef->position().z()) < 304.5) continue; //EB
```
Can't these 123.8 and 304.5 be taken from some definition in dataformats or geometry?
Otherwise, define some constant with reasonable name for these boundaries and use it

rafaellopesdesa replied:
I can't find an easy way to do this. Can it be done like in the TRK-POG code here:

https://github.com/cms-sw/cmssw/blob/master/SimGeneral/TrackingAnalysis/python/TrackingParticleSelection_cfi.py

ie, defined as a parameter?

perrotta replied:
yes, a const parameter with a reasonable name is better than a magic number

perrotta commented on RecoEgamma/EgammaMCTools/plugins/PFClusterMatchedToPhotonsSelector.cc:
```diff
@@ -0,0 +1,114 @@
+      bool isConversion = false;
+      for (auto&& vertRef : trackingParticle.decayVertices()) {
+	if (vertRef->position().rho() > 123.8 && std::abs(vertRef->position().z()) < 304.5) continue; //EB
+	if (std::abs(vertRef->position().z()) > 317.0) continue; // EE
```
Can't this 317.0 be taken from some definition in dataformats or geometry?
Otherwise, define some constant with reasonable name for these boundaries and use it

rafaellopesdesa replied:
Same comment as above. I will put as a parameter, following the selection done in the tracking POG code.

perrotta commented on RecoEgamma/EgammaMCTools/python/pfClusterMatchedToPhotonsSelector_cfi.py:
```diff
@@ -0,0 +1,10 @@
+import FWCore.ParameterSet.Config as cms
+
+particleFlowClusterECALMatchedToPhotons = cms.EDProducer('PFClusterMatchedToPhotonsSelector',
```
This duplicates the config automatically produced by ParameterSetDescription
Please remove this file

rafaellopesdesa replied:
I will remove the description here and leave the python file, if that's ok with you.

perrotta replied:
No: please stick on fillDescription and remove this file.
Explicit python config can be used if you want to clone and modify the default module

rafaellopesdesa committed:
dictionary for genparticles ref valuemap

rafaellopesdesa committed:
code review and add auxiliary products for PS and GEN matching

cmsbuild commented:
Pull request #18106 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

perrotta commented:
Could you please re-implement the fillDescription() method that you removed in your last commit, and consequently remove the pfClusterMatchedToPhotonsSelector_cfi.py (unless you use that file to define a different configuration than the default)?

rafaellopesdesa commented:
@perrotta Yes, I will do. Will update shortly.

rafaellopesdesa committed:
reinstate fillDescription

cmsbuild commented:
Pull request #18106 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

rafaellopesdesa commented:
@perrotta This is a summary of the recent changes to make the review easier:

    * Added EE -> PS PFCluster map. This is also necessary for PF studies.
    * Added PFCluster -> GEN map. This is used for energy calibration of PFClusters.

For the latter, I needed to create the dictionary for edm::ValueMap<reco::GenParticle> ... I hope the code is acceptable. Just pointing out that I added the classes.h and classes_def.xml in this update.

The rest of the code is the same, with the modifications from your code review. Let me know if further changes are needed.

perrotta commented on RecoEgamma/EgammaMCTools/plugins/PFClusterMatchedToPhotonsSelector.cc:
```diff
@@ -0,0 +1,167 @@
+      if (trackingParticle.pdgId() != 22) continue;
+      if (trackingParticle.status() != 1) continue;
+      matchedKey = trackingParticle.genParticles().at(0).key();
+      if (reco::deltaR(trackingParticle, pfCluster.position()) > matchMaxDR_) continue; 
```
For performance purposes recoDeltaR2 must be preferred (and consequently also matchMaxDR_ must transform into matchMaxDR2_)

rafaellopesdesa replied:
@perrotta Ok, I fixed. I also added the DEDR cut that I had the variable in the code but never used. In EGM, we do these studies in particle gun samples, so the matching is quite simple (there is only object in the event), but it is good to have an energy consistency as well to improve the matching against pileup.

perrotta reviewed (commented):
If you can also fix this (sorry, I missed it in the previous review) I will launch the tests

rafaellopesdesa committed:
use DR2 and implement DEDR2

cmsbuild commented:
Pull request #18106 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

rafaellopesdesa committed:
protection for pfcluster energy == 0

cmsbuild commented:
Pull request #18106 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

perrotta commented:
@rafaellopesdesa : do you also plan to move out of the inner loop the calculation of 1./pfCluster.energy() and rename DR2 and DE according to the CMS habit, so that we can run tests, sign off and get rid of this PR?

rafaellopesdesa commented:
@perrotta I have already renamed DR2 and DE in the last commit. I don't see a gain in calculating 1./pfCluster.energy() outside the loop. I can do, but I don't understand why. What would the reason be?

In time: no, an ECAL PFCluster will never have a zero energy. At least the seed will have E>0.

perrotta commented:
I understand this is kind of a private tool. But it is going to enter in the official CMSSW release, and at some point someone could think about using it to do those matches in the official validation sequence, with possibly a huge amount of trackingParticle's.
Now you have one check and one division repeatedly computed with the same outcome for all trackingParticles. If you move it out of that loop, you can do it only once, and just make a much lighter multiplication inside the loop. Not a huge improvement, admitttently. But since we are on it, why not doing it now?

rafaellopesdesa committed:
preselects PFClusters

cmsbuild commented:
Pull request #18106 was updated. @perrotta, @cmsbuild, @slava77, @davidlange6 can you please check and sign again.

rafaellopesdesa commented:
@perrotta Ok, updated. I put a check on the energy outside as a sort of preselection. Also made the loop on PFClusters based on the vector index to avoid having integers floating around, what would be error-prone.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 6a98b9b6b660d3aff7459882eef53225854dcddb
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/18974/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/18974/summary.html

Comparison Summary:
 - You potentially added 17 lines to the logs
 - results differ 1206 all_OldVSNew_TTbarPUwf50202p0 JR results differ 380 all_mini_OldVSNew_TTbarPUwf50202p0 SUMMARY Reco comparison results: 1586 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1913169
 - DQMHistoTests: Total failures: 6179
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1906817
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

perrotta commented:
There is a dictionary error (see "Duplicate Dictionaries" logs in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/18974/summary.html , from https://github.com/cms-sw/cmssw/pull/18106#issuecomment-292190286).

What in lostDefs.txt actually corresponds to wrong dictionary locations. This can also be reproduced (in order to test your fix, when done) with
    duplicateReflexLibrarySearch.py --dir ./ --lostDefs
The new dictionary has just GenParticle. It should then be moved to HepMCCandidate in DataFormats/HepMCCandidate/classes* files
(Thank you @slava77  for the explanations!)

Please move the new dictionaries to that other package.

rafaellopesdesa commented:
Well, I also learned something new today. I will do the fix and test...
Then update. Thanks to you both for the explanation.

On Apr 6, 2017 6:27 PM, "perrotta" <notifications@github.com> wrote:

There is a dictionary error (see "Duplicate Dictionaries" logs in
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-
request-integration/PR-18106/18974/summary.html , from #18106 (comment)
<https://github.com/cms-sw/cmssw/pull/18106#issuecomment-292190286>).

What in lostDefs.txt actually corresponds to wrong dictionary locations.
This can also be reproduced (in order to test your fix, when done) with
duplicateReflexLibrarySearch.py --dir ./ --lostDefs
The new dictionary has just GenParticle. It should then be moved to
HepMCCandidate in DataFormats/HepMCCandidate/classes* files
(Thank you @slava77 <https://github.com/slava77> for the explanations!)

Please move the new dictionaries to that other package.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cms-sw/cmssw/pull/18106#issuecomment-292229110>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AFtCpCmvdVpK48ySjrNGvqLKWM_SozZmks5rtRJLgaJpZM4MreZx>
.

rafaellopesdesa committed:
move dictionary to DataFormats.HepMCCandidate

rafaellopesdesa commented:
I've updated and tested with the commands here:
https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Checking_the_Software

From the command you suggested, the only output I see is:

**** SKIPPING './CondCore/Utilities/src/classes_def.xml' - Doesn't seem to have proper information.

Which is not related to my PR. I have also tested with the "limited" runTheMatrix and the only errors I see are annoying DAS errors. I will push now.

Thanks again.

cmsbuild commented:
Pull request #18106 was updated. @perrotta, @perrozzi, @thuer, @cmsbuild, @slava77, @govoni, @davidlange6 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 5638c70947fe42c677378f02e80f96b2b25ef1fe
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/19020/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18106/19020/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1917345
 - DQMHistoTests: Total failures: 117
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1917055
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

perrotta commented:
+1
New producer used in some local test, with no effext on standard workflows
Implementation looks correct

rafaellopesdesa commented:
@perrozzi, @thuer, @govoni In order to help the generator approval, I would like to describe the single change I did in DataFormats/HepMCCandidate. I added to the dictionaries generated:

   * edm::ValueMap < reco::GenParticleRef >
   * edm::Wrapper < edm::ValueMap < reco::GenParticleRef > > 

Nothing else was changed.

perrozzi commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @muzaffar, @davidlange6, @smuzaffar

davidlange6 commented:
+1
